Classfier: add a history manager to workflow steps

eatyourgreens opened the pull request:
This adds [`UndoManager`](https://github.com/mobxjs/mobx-state-tree/blob/master/packages/mst-middlewares/README.md#undomanager) to the tasks area, recording the history of the task area state (the current step and its annotations.) 'Next' is wired up to push the active step onto the history stack before selecting the next step. 'Back' is wired up to use `undoManager.undo` to move back one step in the history stack. `getPreviousStepKey` and `isThereAPreviousStep` are removed because we now use `undoManager.canUndo` to check if we can go back and `undoManager.latest` to get the previous history entry.
https://localhost:8080/?project=12268&workflow=17076&env=production&demo=true

Recursive workflows aren't supported yet. I'm going to add support for those in a follow-up PR.

TODO:
- [x] test the new models.

Package:
lib-classifier

Closes #1781.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented:
The next step, I think, is using the history stack to populate the Tasks component. That will allow us to support recursive workflows.

eatyourgreens commented:
Also, `UndoManager` supports [clearing just the redo stack](https://github.com/mobxjs/mobx-state-tree/blob/67ca1a28b55de2baed689c7f5e048182b6f9a36e/packages/mst-middlewares/src/undo-manager.ts#L210-L214), which we can use when you go back to a branching step and choose a different branch.

eatyourgreens commented:
The classifier task stories are still working on this branch, so that's good.

srallen commented:
@eatyourgreens is this ready for review? I see a TODO comment in the original post. I'd like to have a look at this to see if it happens to resolve a bug with the transcription task. Thanks!

eatyourgreens commented:
It needs tests, but otherwise I think it should work. I haven't tested it with any workflows that use feedback.

srallen reviewed (commented):
This fixes the back button display for the transcription task:

<img width="422" alt="Screen Shot 2021-03-01 at 11 49 49 AM" src="https://user-images.githubusercontent.com/5016731/109538090-4c0b9a80-7a85-11eb-9023-4cb7b80e32bd.png">

I've only done a quick read through of the code, not a thorough review, so I'll just leave a comment review. I have a conceptual question: Why create a new `AnnotatedSteps` and `AnnotatedStep` model instead of refactoring `WorkflowSteps` and `Step` to have this functionality? The purpose of `WorkflowSteps` and `Step` is to store and manage the next/back step flow of the workflow. It seems `AnnotatedSteps` and `AnnotatedStep` does the same thing but with the undo manager. Is the intention that the new models replaces these or not? How do they relate to each other if it's not a replacement? It's a bit confusing to have both.

eatyourgreens commented:
I've added some tests, which uncovered a bug in the `reset` action but hopefully also helps explain what the history manager is doing: recording steps prior to the active workflow step.
```
 Model > AnnotatedSteps
Fetching more subjects
    ✓ should exist
    when the workflow starts
      ✓ should be empty
      ✓ should not have a latest step
      ✓ should not be able to undo
    after the first workflow step
      ✓ should have one step
      ✓ should store the first workflow step
      ✓ should store the first step's annotations
      ✓ should be able to undo
      on undo
        ✓ should be empty
        ✓ should not have a latest step
        ✓ should not be able to undo
    on reset
      ✓ should be empty
      ✓ should not have a latest step
      ✓ should not be able to undo

```

srallen commented:
@eatyourgreens thanks for the tests, but could you explain why your approach is adding another store rather than refactoring the existing steps stores? How do the new stores relate in function to the existing stores and what the existing `Step` store is now doing with its actions for setting the next and previous step if the history manager takes over handling the flow of the steps? What I'm trying to avoid here is a confusing state of partial refactoring and cruft creation that someone new to this would have difficulty putting together.

eatyourgreens commented:
The history manager isn't handling the flow of steps. `WorkflowSteps` still handles selecting next/previous steps. The history records a stack of steps you've already completed, that's all.

srallen commented:
The functionality is mostly duplicative essentially. The `Step` store is also recording what the previous step was. I'm supportive of using the undo manager middleware, but we don't need both. We do still need the logic to determine if there's branching and which step is next depending on the branched choice, but that's it? The undo manager can handle the rest, couldn't it? So, with that in mind, the undo manager could be added to `Step` directly and replace the `setPrevious` action?

eatyourgreens commented:
I did have a look at how the previous/next steps are being set, but couldn't figure it out. There does seem to be some hidden magic going on, in that `workflowSteps.selectStep()` automatically selects the next step in a workflow, but it's not obvious how that works.

The code here does use `step.next` to select the next step in a workflow. I think `step.previous` can perhaps be removed?

eatyourgreens commented:
For recursive workflows, we need to be able to store each step more than once. That's covered in #2050 but I don't see any way to do that with a refactor. I think it means adding new functionality that's not currently supported in the store.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/components/TaskNavButtons/TaskNavButtonsContainer.js:
```diff
@@ -1,37 +1,40 @@
 
-@inject(storeMapper)
-@observer
 class TaskNavButtonsContainer extends React.Component {
```
#782

srallen commented:
Yes, at the very least `step.previous` and `step.setPrevious` is now cruft with this change. I think what would help is if you could diagram the relationships between the new stores and annotations with this change and the change in #2050. It might implement the feature asked for, but making sense of it in the actual code is really difficult to do now. We have a state spider web not a state tree at this point. In #2050, the source of truth for the state of annotations now appears to belong to the `AnnotatedSteps` store, not classifications or tasks. Or is it duplicated? Does this mean now there's cruft with the task stores now? I'm not really sure anymore.

eatyourgreens commented:
`AnnotatedSteps` as the source of truth for annotations is a really good idea. That would solve some problems with keeping branching workflows, like TNA, in sync with the classification store. I might try that on Monday: store annotations directly in history, rather than references. 

It may not work, in that UndoManager records every action in the sub tree, and we don’t want to record `annotation.update()`. I’ve already found that `createAnnotation()` can’t be undone and redone, because it generates identifier mismatches in the classification model.

eatyourgreens commented:
I've removed `step.previous` and `step.setPrevious()`. I've also changed the model to use the declarative syntax for `UndoManager`, so the history manager is now private.

eatyourgreens commented:
I'm wondering whether to close this PR in favoour of #2050, which includes this branch and adds support for branching, recursive workflows.

srallen commented:
@eatyourgreens the changes in this PR don't fully make sense without the context of #2050, so I'm supportive of that even though it makes it a very large PR to review.

eatyourgreens commented:
@srallen Thanks! I'm hoping that a lot of the new code is tests but, as @shaunanoordin commented on 32050, there are a lot of moving parts here, which makes me nervous.

Shaun found some bugs in the branching step logic too, which are triggered once a branching question can be answered more than once. I've fixed those on #2050 rather than here.

I'll close this but won't delete the branch, in case it turns out we want to refer back to it.
